File: leetcode/Easy/0141-linked-list-cycle/0141-linked-list-cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if(head == NULL)
            return false;
        ListNode* one = head;
        ListNode* two = head->next;
        while (one != two){
            if (two == NULL || two->next == NULL)
                return false;
            one = one->next;
            two = two->next->next;
        }
        return true;
    }
};

File: leetcode/Easy/0141-linked-list-cycle/0141-linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if not head:
            return False
        one = head
        two = head.next
        while one != two:
            if not two or not two.next:
                return False
            one = one.next
            two = two.next.next
        return True
